src/iop/sigmoid.c

```diff
@@ -0,0 +1,631 @@
+  const float ref_film_fog = 0.0f;
+  const float ref_paper_exposure = powf(ref_film_fog + MIDDLE_GREY, ref_film_power) * ((ref_magnitude / MIDDLE_GREY) - 1.0f);
+  const float delta = 1e-6f;
+  const float ref_slope = (generalized_loglogistic_sigmoid(MIDDLE_GREY + delta, ref_magnitude, ref_paper_exposure, ref_film_fog, ref_film_power, ref_paper_power) -
+                           generalized_loglogistic_sigmoid(MIDDLE_GREY - delta, ref_magnitude, ref_paper_exposure, ref_film_fog, ref_film_power, ref_paper_power)) / 2.0f / delta;
+
+  // Add skew
+  module_data->paper_power = powf(5.0f, -params->contrast_skewness);
+
+  // Slope at low film power
+  const float temp_film_power = 1.0f;
+  const float temp_white_target = 0.01f * params->display_white_target;
+  const float temp_white_grey_relation = powf(temp_white_target / MIDDLE_GREY, 1.0f / module_data->paper_power) - 1.0f;
+  const float temp_paper_exposure = powf(MIDDLE_GREY, temp_film_power) * temp_white_grey_relation;
+  const float temp_slope = (generalized_loglogistic_sigmoid(MIDDLE_GREY + delta, temp_white_target, temp_paper_exposure, ref_film_fog, temp_film_power, module_data->paper_power) -
+                            generalized_loglogistic_sigmoid(MIDDLE_GREY - delta, temp_white_target, temp_paper_exposure, ref_film_fog, temp_film_power, module_data->paper_power)) / 2.0f / delta;
+
+  // Figure out what film power fulfills the target slope
+  // (linear when assuming display_black = 0.0)
+  module_data->film_power = ref_slope / temp_slope;
+
+  // Calculate the other parameters now that both film and paper power is known
+  module_data->white_target = 0.01f * params->display_white_target;
+  module_data->black_target = 0.01f * params->display_black_target;
+  const float white_grey_relation = powf(module_data->white_target / MIDDLE_GREY, 1.0f / module_data->paper_power) - 1.0f;
+  const float white_black_relation = powf(module_data->black_target / module_data->white_target, -1.0f / module_data->paper_power) - 1.0f;
+
+  module_data->film_fog = MIDDLE_GREY * powf(white_grey_relation, 1.0f / module_data->film_power) / (powf(white_black_relation, 1.0f / module_data->film_power) - powf(white_grey_relation, 1.0f / module_data->film_power));
+  module_data->paper_exposure = powf(module_data->film_fog + MIDDLE_GREY, module_data->film_power) * white_grey_relation;
+
+  module_data->color_processing = params->color_processing;
+  module_data->hue_preservation = fminf(fmaxf(0.01f * params->hue_preservation, 0.0f), 1.0f);
+}
+
+#ifdef _OPENMP
+#pragma omp declare simd
+#endif
+static inline void desaturate_negative_values(const float pix_in[4], float pix_out[4])
+{
+  const float pixel_average = fmaxf((pix_in[0] + pix_in[1] + pix_in[2]) / 3.0f, 0.0f);
+  const float min_value = fminf(fminf(pix_in[0], pix_in[1]), pix_in[2]);
+  const float saturation_factor = min_value < 0.0f ? -pixel_average / (min_value - pixel_average) : 1.0f;
+  for(size_t c = 0; c < 3; c++)
```

Comment:
After converting to use `dt_aligned_pixel_t`, consider using the `for_each_channel` macro for vectorizability. There are a plenty of examples of its usage in the codebase.